### ISSUE TITLE
Allow incremental annotation processors to write resources

### ISSUE THREAD

**Issue by oehme** [gradle/gradle#4702]
For instance a processor might want to generate a database schema based on annotated entity classes.

Supporting `Filer.createResourceFile` wouldn't be a big problem. But as far as I've seen, most processors that create resources don't use the `Filer`, because they don't want to mix generated resources with generated .java files. Instead they ask users to provide a processor argument, pointing the processor to a different location and then they use `java.io` directly. This means that Gradle will miss the creation of these files and thus would not clean them up when necessary.

Ideally the Filer API would provide a way for processors to define new output locations, which the user can then configure using compiler arguments. Right now only `source output` and `class output` are supported.

**Comment by cesar1000** [comment]
Incidentally, we have an annotation process that generates a database schema into the build directory, and we add it to the compiler task explicitly as an input.

We also use AutoService on our annotation processors, in order to publish the processor for discovery by the compiler. It looks like it uses the `Filer` API properly: https://github.com/google/auto/blob/master/service/src/main/java/com/google/auto/service/processor/AutoServiceProcessor.java

**Comment by oehme** [comment]
> we add it to the compiler task explicitly as an input.

You mean as an output directory?

**Comment by cesar1000** [comment]
Yeah, sorry, we make it an explicit output in order to enable caching:

```
    variant.javaCompile.with {
        // Declare the schema output directory as a compilation output so that it's stored in the cache.
        outputs.dir(schemaOutputDirectory)
                .withPropertyName('schemaOutputDirectory')
    }
```

**Comment by hzsweers** [comment]
Another example: we just open sourced this project for breadcrumbing metadata across compile-time boundaries. We do it by generating the data into resources and then consuming them in consumers of the library.

**Comment by oehme** [comment]
@hzsweers I'm afraid that processor won't work incrementally for deeper reasons: It's accessing internals of the compiler (e.g. the file manager) to get those resources. So we would have a hard time tracking its inputs and outputs. The casts it does to get those internals would also fail since we decorate the processing environment. 

I think to make this work we'd ideally want an official addition to the processor API to load resources from the classpath. Otherwise we'd need some out of band channel for you to communicate to Gradle what you are doing. I'm open for other ideas of course.

**Comment by hzsweers** [comment]
Not being able to operate incrementally when reading data makes sense, but I don't see why producing data can't be incremental since it only writes data through normal file object APIs?

**Comment by cesar1000** [comment]
Also, would it make sense to use gradle variants for passing resources
downstream instead of Java resources (when running in gradle)? You should
be able to write a gradle plugin to set up the annotation processor
parameters and add inputs and outputs to the compiler task.

**Comment by oehme** [comment]
@hzsweers The writing part is not a big problem, but that's only half of your processor. I was just pointing out that fixing this won't be enough for your use case.

**Comment by hzsweers** [comment]
Any update on this? This is actually going to break Dagger once Android's R8 tool is finished, as it generates R8 .pro files into resources. I was talking with Jerome from the android gradle plugin team at the android dev summit as well about this, and he confirmed this causes full recompilations with dagger in gradle 5.0

**Comment by isker** [comment]
@oehme would it be acceptable to start with just supporting `Filer.createResourceFile`?  I am willing to take a crack at it if you think it's reasonable.

**Comment by oehme** [comment]
@isker yes, that would be a start. You'd have to add resource tracking to our incremental compile infrastructure, as we currently only track which classes to recompile.

**Comment by bingranl** [comment]
Lifecycle use `Filer.createResource` generate keep rule .pro files and that is the only remaining issue to make it isolating AP. Any plans for adding support `Filer.createResource`? As far as I know, kapt is adding incremental support for `Filer.createResource`.

**Comment by oehme** [comment]
@isker is currently working on that

**Comment by isker** [comment]
Sorry it’s been so slow. I haven’t had much time. I am finishing up testing
and should have a PR soon.

**Comment by oehme** [comment]
Thanks to @isker, `Filer#createResource` is now supported. 

Note that writing files any other way (e.g. using `java.io.File`) is still *not* supported. I don't see a good way to support that, since the compiler itself provides no API to define new output locations.

**Comment by bingranl** [comment]
Awesome! Thanks for this great PR @isker and including it in 5.4 @oehme !

**Comment by ZacSweers** [comment]
Two questions on the above solution:
* Should a processor that consumes elements from transitive dependencies indicated via resources, should those elements be marked as originating elements even though they're not in the project?
* In the event of dependency changes, will the next build always run non-incrementally?

**Comment by oehme** [comment]
> Should a processor that consumes elements from transitive dependencies indicated via resources, should those elements be marked as originating elements even though they're not in the project?

There is no way in the processor API to read resources from upstream projects. You can only read them from the current sources. 

But since you're asking I assume you've hacked into javac internals to get to the underlying JARs. 
Depending on which path you look at, this will either be broken or lead to full recompiles. If you look on the classpath, you won't be rerun on changes, because Gradle ignores resources on the classpath. If you look on the processor path, it'll work, but any processor path change leads to a full recompile, so you won't benefit from incremental processing.

> In the event of dependency changes, will the next build always run non-incrementally?

Depends on what dependency. Classpath changes are done incrementally (we do JAR diffs). Processor path changes lead to a full recompile.

**Comment by ZacSweers** [comment]
I don't think it's very hacky! https://github.com/uber/crumb/blob/master/crumb-core/src/main/kotlin/com/uber/crumb/core/CrumbManager.kt#L65. We're just reading out of the current classloader provided by the ProcessingEnvironment, which I think falls under your comment about looking on the processor path.

> There is no way in the processor API to read resources from upstream projects

Just to clarify - if project A depends on project B, I mean that A wants to read resources from B's jar. Is that what you mean by upstream? Asking since my mental model is that A is upstream from B

**Comment by oehme** [comment]
`JavacProcessingEnvironment` is an internal class that processors are not supposed to call. From the documentation:

```
This is NOT part of any supported API. If you write code that depends on this, you do so at your own risk. This code and its internal interfaces are subject to change or deletion without notice.
```

> We're just reading out of the current classloader provided by the ProcessingEnvironment, which I think falls under your comment about looking on the processor path.

You're reading from the [classpath](https://github.com/uber/crumb/blob/master/crumb-core/src/main/kotlin/com/uber/crumb/core/CrumbManager.kt#L67), which is broken for the reasons explained above.

>  if project A depends on project B, I mean that A wants to read resources from B's jar.

I understand, this is not allowed by the java compiler API. Your processor is operating outside the processing API, so there is no guarantee it'll work correctly.

**Comment by ZacSweers** [comment]
I see. Let's consider a different, more practical example: SPIs. If project B produced a service file in resources that project A tried to consumer - is it not possible then for that to be done in a way that adhere's to gradle's incremental processing paradigm?

**Comment by oehme** [comment]
You can't access resources from other projects, that's simply not part of the compiler API. You can only see classes from upstream projects.

**Comment by ZacSweers** [comment]
> It's just not allowed by the Java compiler API

Considering this is how service loading is implemented, why is that not allowed?

**Comment by oehme** [comment]
Service loading is a runtime concept, not a compile time concept. There are many other things that the compiler API does not give you access to, like method bodies of upstream classes, arbitrary file IO etc. 

These are very important limitations that allow build systems to be more incremental and parallel. Gradle's incremental compiler is one example. Another would be the Eclipse compiler or Bazel's ABI jar mechanism. A processor that works around those limitations will not work correctly in any of those incremental environments.

I highly recommend considering a different (incremental-) compiler friendly approach. The [Dagger processor](https://github.com/google/dagger) is a great example of that. It uses `@Module`-annotated classes to aggregate bindings. As a result, it doesn't require any classpath scanning or resource access and works really well in incremental builds.

**Comment by thyming** [comment]
Is marking an annotation processor using this feature backwards compatible? I.e. if an annotation processor that produces resources is marked as incremental, would there be any adverse effects for users of gradle with versions less than 5.4?

**Comment by oehme** [comment]
No, there wouldn't be any problems. It will just be treated as non-incremental in older versions.

**Comment by thyming** [comment]
Thank you, @oehme.
